{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Getting Started with ADAM and the NGS: A Walk Through the Schemas of ADAM\n",
      "\n",
      "ADAM is a genome processing engine and set of associated file formats meant to standarize and scale the processing of the the massive amounts of data generated by next generation sequencing (NGS). The most contribution of ADAM are the standard and powerful representations of NGS data defined by it via its Avro schemas. This allows for much greater flexibility in terms of the tools which can be used to process genomic data.\n",
      "\n",
      "Avro is useful because it can be used with e.g. Parquet to distribute these data across a cluster, as well as because there are adapters written for many languages which transform Avro records into objects native to the language. \n",
      "\n",
      "### Loading the Schemas\n",
      "[ADAM's primary records](https://github.com/bigdatagenomics/adam/blob/master/adam-format/src/main/resources/avro/adam.avdl) are described using the [Avro interface description language](http://avro.apache.org/docs/1.7.5/idl.html). In order to be parsed by Python's Avro library, the IDL representation must be transformed to the standard JSON Avro Protocol representation using avro-tools [(described here)](http://avro.apache.org/docs/1.7.5/gettingstartedjava.html). \n",
      "\n",
      "Once this is done, the protocol may be loaded and parsed by the Python library. The API isn't documented (well, it may be, and probably is, but I cannot find it), but the code is easy to read. We want to inspect the types (which will include records) that ADAM uses.\n",
      "\n",
      "The IDL representation is much easier to read than the JSON representation, so we'll be looking at the IDL while working with the \"raw\" JSON."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import avro.protocol as apr"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "protocol = apr.parse(open(\"adam.avpr\", \"r\").read()) \n",
      "[tp for tp in protocol.types_dict if protocol.types_dict[tp].type == 'record'] # we're ignoring enums and other schema types"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Reading and Writing Schemas\n",
      "\n",
      "We need to be able to read and write records for a given schema from and to disk. We will demonstrate that for an arbitrary record (a mock ADAMContig) below:"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from avro.datafile import DataFileReader, DataFileWriter\n",
      "from avro.io import DatumReader, DatumWriter"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "schema = protocol.types_dict['ADAMContig']\n",
      "schema.to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# this is the Python representation of the Avro object. \n",
      "ADAM_contig = {'contigId':9230, 'referenceURL':'http://data.dna/223', 'contigName':\"1Nabc\", \n",
      "               'contigLength':7781, 'contigMD5':\"8743b52063cd84097a65d1633f5c74f5\"} # mock data"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "with DataFileWriter(open(\"contigs.avro\", \"w\"), DatumWriter(), schema) as contig_writer:\n",
      "    contig_writer.append(ADAM_contig)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "with DataFileReader(open(\"contigs.avro\", \"r\"), DatumReader()) as contigs:\n",
      "    for contig in contigs:\n",
      "        print contig"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# The Schemas Themselves\n",
      "We can inspect the JSON schema of the record, and see that which fields (and their respective types) that record contains.\n",
      "\n",
      "We'll write a few records to a new file, using some arbitrary data (but of the correct type), and then load that file and print the Python representations of those records. This is about all there is to do with these records."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMRecord\n",
      "\n",
      "An ADAMRecord stores alignment against a reference along with quality data. It is the ADAM equivalent to a SAM/BAM file (but much easier to read).\n",
      "\n",
      "Below is a real example of a real ADAMRecord, exported to JSON (which happily happens to have the same syntax as Python dictionaries). This was generated by converting an existing BAM file to ADAM format (e.g. `adam bam2adam hg37.bam hg37.bam.adam`), and then printing it (into JSON) with e.g. `adam print hg37.bam.adam`.\n",
      "\n",
      "A Record, like a BAM file, stores a read sequence \\[sequence\\] with quality \\[qual\\] data (ASCII+33 + [PHRED score](http://en.wikipedia.org/wiki/Phred_quality_score)), along with alignment information stored in extended CIGAR format [cigar]. More information on all fields can be found in the comments on the IDL representation below.\n",
      "\n",
      "Additionally, important metainformation is parsed out of the BAM files and stored as first-class key-vals. Nonstandard additional attributes are stored in a string under `attributes`."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ADAM_record = {\"referenceName\": \"20\", \"referenceId\": 19, \"start\": 19893804, \"mapq\": 60, \n",
      "               \"readName\": \"20GAVAAXX100126:4:64:6132:191287\", \n",
      "               \"sequence\": \"GTTTTCTATGAAGTTATTTTCTAGGGATTCTGTTTTGTTGTCGTTGTTCACACTGTAGCTCTCAGATCTTACTGTTTTTTTTTTAATTGTGATAAAGCATA\", \n",
      "               \"mateReference\": \"20\", \"mateAlignmentStart\": 19893476, \"cigar\": \"101M\", \n",
      "               \"qual\": \"EHHHFGDGHFHF8EEFB=B@=GHFFBAA@8??>IHHEIHEH>EHH@HFG>GEGHEGHFA<GDGHFFHGHGFCFCE@EEHHHHHGGFHGEGFFFGGGFBDDD\", \n",
      "               \"recordGroupName\": \"20GAV.4\", \"recordGroupId\": 11, \"readPaired\": true, \"properPair\": true, \"readMapped\": true, \n",
      "               \"mateMapped\": true, \"readNegativeStrand\": true, \"mateNegativeStrand\": false, \"firstOfPair\": true, \n",
      "               \"secondOfPair\": false, \"primaryAlignment\": true, \"failedVendorQualityChecks\": false, \n",
      "               \"duplicateRead\": false, \"mismatchingPositions\": \"101\", \n",
      "               \"attributes\": \"XT:A:U\\tOQ:Z:HHHHFHGHHHHG=FGGC?DC@HHHHFDDA:?E>HHHHHHHHHHHHFHGHGHGHHHHHHF@HFHHHHHHHHHGGGGGGGHHHHHHHHHHHHHHHHHHHGHHH\\tMQ:i:60\\tBQ:Z:@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@\\tXO:i:0\\tXM:i:0\\tSM:i:37\\tNM:i:0\\tAM:i:37\\tXG:i:0\\tRG:Z:20GAV.4\\tX1:i:0\\tX0:i:1\", \n",
      "               \"recordGroupSequencingCenter\": \"BI\", \"recordGroupDescription\": null, \"recordGroupRunDateEpoch\": null, \n",
      "               \"recordGroupFlowOrder\": null, \"recordGroupKeySequence\": null, \"recordGroupLibrary\": \"Solexa-18484\", \n",
      "               \"recordGroupPredictedMedianInsertSize\": null, \"recordGroupPlatform\": \"illumina\", \"recordGroupPlatformUnit\": \"20GAVAAXX100126.4\", \n",
      "               \"recordGroupSample\": \"NA12878\", \n",
      "               \"mateReferenceId\": 19, \"referenceLength\": 63025520, \"referenceUrl\": null, \n",
      "               \"mateReferenceLength\": 63025520, \"mateReferenceUrl\": null, \"origQual\": null}\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Below is the record from the SAM file that is represented above:"
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "20GAVAAXX100126:4:64:6132:191287     83     20     19893805     60     101M     =     19893477     -428    GTTTTCTATGAAGTTATTTTCTAGGGATTCTGTTTTGTTGTCGTTGTTCACACTGTAGCTCTCAGATCTTACTGTTTTTTTTTTAATTGTGATAAAGCATA   \n",
      "EHHHFGDGHFHF8EEFB=B@=GHFFBAA@8??>IHHEIHEH>EHH@HFG>GEGHEGHFA<GDGHFFHGHGFCFCE@EEHHHHHGGFHGEGFFFGGGFBDDD   \n",
      "X0:i:1  X1:i:0  MD:Z:101  RG:Z:20GAV.4  XG:i:0  AM:i:37  NM:i:0  SM:i:37  XM:i:0  XO:i:0  MQ:i:60  XT:A:U BQ:Z:@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The IDL for ADAMRecord is below, with all fields commented."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "```c\n",
      "record ADAMRecord {\n",
      "\n",
      "    //////////////////////////////////\n",
      "    // Primary alignment fields     //\n",
      "    //////////////////////////////////\n",
      "    // This is the Query Template Name (the name of the read of the same template)\n",
      "    //  Corresponds to QNAME (column 1) in SAM.\n",
      "    union { null, string } readName = null;\n",
      "    // The reference sequence name of the alignment.\n",
      "    // Corresponds to RNAME (column 3) in SAM.\n",
      "    union { null, string } referenceName = null;\n",
      "    // Reference position start of the first matching base.\n",
      "    // Corresponds to POS (column 4) in SAM (adjusted to be 0-indexed).\n",
      "    union { null, long } start = null;\n",
      "    // The PHRED-scored mapping quality \n",
      "    // -10Log_10 Pr{the mapping position is wrong}.\n",
      "    // Corresponds to MAPQ (column 5) in SAM.\n",
      "    union { null, int } mapq = null;\n",
      "    // CIGAR string, indicated alignment, indels, misalignment to reference\n",
      "    // Corresponds to CIGAR (column 6) in SAM.\n",
      "    union { null, string } cigar = null;\n",
      "    // The referenceName of the mate template (the next read in the template).\n",
      "    // Corresponds to RNEXT (column 7) in SAM.\n",
      "    union { null, string } mateReference = null;\n",
      "    // The position of the primary alignment of the next read in the template.\n",
      "    // Corresponds to PNEXT (column 8) in SAM (adjusted to be 0-indexed).\n",
      "    union { null, long } mateAlignmentStart = null;\n",
      "    // ASCII+33 Phred scores for the bases of sequence.\n",
      "    // Corresponds to SEQ (column 10) in SAM.\n",
      "    union { null, string } qual = null;\n",
      "    // The actual sequence of bases. \n",
      "    // Corresponds to QUAL (column 11) in SAM.\n",
      "    union { null, string } sequence = null;\n",
      "    \n",
      "    \n",
      "    //////////////////////////////////////////\n",
      "    // Read flags (all default to false)    //\n",
      "    // Derived from FLAG (column 2) in SAM  //\n",
      "    //////////////////////////////////////////\n",
      "    union { boolean, null } readPaired = false;          // the template has multiple segments (0x1 in SAM) \n",
      "    union { boolean, null } properPair = false;          // each segment is properly aligned (0x2 in SAM)\n",
      "    union { boolean, null } readMapped = false;          // the segment is mapped (opposite of 0x4 in SAM)\n",
      "    union { boolean, null } mateMapped = false;          // the mate is mapped (opposite of 0x8 in SAM)\n",
      "    union { boolean, null } readNegativeStrand = false;  // sequence is reverse complimented (0x10 in SAM)\n",
      "    union { boolean, null } mateNegativeStrand = false;  // mate is reverse complimented (0x20 in SAM)\n",
      "    union { boolean, null } firstOfPair = false;         // first segment in the template (0x40 in SAM)\n",
      "    union { boolean, null } primaryAlignment = false;    // the first segment in the template (0x40 in SAM)\n",
      "    union { boolean, null } secondOfPair = false;        // the last segment in the template (0x80 in SAM)\n",
      "    union { boolean, null } failedVendorQualityChecks = false;  // not passing quality controls (0x200 in SAM)\n",
      "    union { boolean, null } duplicateRead = false;       // PCR or optimal duplicate (0x400 in SAM(\n",
      "    \n",
      "    \n",
      "    //////////////////////////////\n",
      "    // Derived & header fields  //\n",
      "    //////////////////////////////\n",
      "    // The corresponding reference length (LN in the SAM header).\n",
      "    union { null, long }   referenceLength = null;\n",
      "    // The corresponding reference URL (UR in the SAM header [NOTE: technically the URI in SAM spec]).\n",
      "    union { null, string } referenceUrl = null;\n",
      "    // The index of the referenceName (derived from the SAM header).\n",
      "    union { null, int } referenceId = null;\n",
      "    // c.f referenceLength\n",
      "    union { null, long } mateReferenceLength = null;\n",
      "    // c.f referenceUrl\n",
      "    union { null, string } mateReferenceUrl = null;\n",
      "    // c.f. referenceId\n",
      "    union { null, int } mateReferenceId = null;\n",
      "    \n",
      "    // This is the name of the record group for the alignment.\n",
      "    // It is taken from the header of the SAM file, by joining \n",
      "    // part of the QNAME with the PU, and taking the associated RG ID.\n",
      "    union { null, string } recordGroupName = null;\n",
      "    // The index of the readGroup.\n",
      "    // (derived from the reference in the SAM header).\n",
      "    union { null, int } recordGroupId = null;\n",
      "\n",
      "\n",
      "    ///////////////////////////////////////\n",
      "    // Commonly used optional attributes //\n",
      "    ///////////////////////////////////////\n",
      "    // String for mismatching positions.\n",
      "    // Taken from the alignment optional field tag MD.\n",
      "    union { null, string } mismatchingPositions = null;\n",
      "    // Original base quality (generally from before BQSR).\n",
      "    // Taken from the alignment optional field tag OQ.\n",
      "    union { null, string } origQual = null;\n",
      "    // Corresponds to column 11 in SAM, less the above optional\n",
      "    // attributes which have been parsed out.\n",
      "    union { null, string } attributes = null;\n",
      "\n",
      "\n",
      "    /////////////////////////////////////////////////\n",
      "    // Record group identifer from sequencing run  //\n",
      "    /////////////////////////////////////////////////\n",
      "    // The sequencing center which produced the read.\n",
      "    // Taken from the CN attribute of the reference group (RG) line in the\n",
      "    // SAM header corresponding to the particular reference of the alignment.\n",
      "    union { null, string } recordGroupSequencingCenter = null;\n",
      "    union { null, string } recordGroupDescription = null;\n",
      "    union { null, long } recordGroupRunDateEpoch = null;\n",
      "    union { null, string } recordGroupFlowOrder = null;\n",
      "    union { null, string } recordGroupKeySequence = null;\n",
      "    union { null, string } recordGroupLibrary = null;\n",
      "    union { null, int } recordGroupPredictedMedianInsertSize = null;\n",
      "    union { null, string } recordGroupPlatform = null;\n",
      "    union { null, string } recordGroupPlatformUnit = null;\n",
      "    union { null, string } recordGroupSample = null;\n",
      "}\n",
      "```"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "With a corresponding ADAM schema found in the protocol dictionary we created earlier (these will henceforth be omitted in favor of IDL):"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "protocol.types_dict['ADAMRecord'].to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "{'fields': [{'default': None,\n",
        "   'doc': u'* These two fields, along with the two\\n     * reference{Length, Url} fields at the bottom\\n     * of the schema, collectively form the contents\\n     * of the Sequence Dictionary embedded in the these\\n     * records from the BAM / SAM itself.\\n     * TODO: this should be moved to ADAMContig',\n",
        "   'name': u'referenceName',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'referenceId', 'type': [u'null', u'int']},\n",
        "  {'default': None, 'name': u'start', 'type': [u'null', u'long']},\n",
        "  {'default': None, 'name': u'mapq', 'type': [u'null', u'int']},\n",
        "  {'default': None, 'name': u'readName', 'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'sequence', 'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'mateReference', 'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'mateAlignmentStart', 'type': [u'null', u'long']},\n",
        "  {'default': None, 'name': u'cigar', 'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'qual', 'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'recordGroupName', 'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'recordGroupId', 'type': [u'null', u'int']},\n",
        "  {'default': False, 'name': u'readPaired', 'type': [u'boolean', u'null']},\n",
        "  {'default': False, 'name': u'properPair', 'type': [u'boolean', u'null']},\n",
        "  {'default': False, 'name': u'readMapped', 'type': [u'boolean', u'null']},\n",
        "  {'default': False, 'name': u'mateMapped', 'type': [u'boolean', u'null']},\n",
        "  {'default': False,\n",
        "   'name': u'readNegativeStrand',\n",
        "   'type': [u'boolean', u'null']},\n",
        "  {'default': False,\n",
        "   'name': u'mateNegativeStrand',\n",
        "   'type': [u'boolean', u'null']},\n",
        "  {'default': False, 'name': u'firstOfPair', 'type': [u'boolean', u'null']},\n",
        "  {'default': False, 'name': u'secondOfPair', 'type': [u'boolean', u'null']},\n",
        "  {'default': False,\n",
        "   'name': u'primaryAlignment',\n",
        "   'type': [u'boolean', u'null']},\n",
        "  {'default': False,\n",
        "   'name': u'failedVendorQualityChecks',\n",
        "   'type': [u'boolean', u'null']},\n",
        "  {'default': False, 'name': u'duplicateRead', 'type': [u'boolean', u'null']},\n",
        "  {'default': None,\n",
        "   'name': u'mismatchingPositions',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'attributes', 'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupSequencingCenter',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupDescription',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupRunDateEpoch',\n",
        "   'type': [u'null', u'long']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupFlowOrder',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupKeySequence',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupLibrary',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupPredictedMedianInsertSize',\n",
        "   'type': [u'null', u'int']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupPlatform',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupPlatformUnit',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'recordGroupSample',\n",
        "   'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'mateReferenceId', 'type': [u'null', u'int']},\n",
        "  {'default': None, 'name': u'referenceLength', 'type': [u'null', u'long']},\n",
        "  {'default': None, 'name': u'referenceUrl', 'type': [u'null', u'string']},\n",
        "  {'default': None,\n",
        "   'name': u'mateReferenceLength',\n",
        "   'type': [u'null', u'long']},\n",
        "  {'default': None, 'name': u'mateReferenceUrl', 'type': [u'null', u'string']},\n",
        "  {'default': None, 'name': u'origQual', 'type': [u'null', u'string']}],\n",
        " 'name': u'ADAMRecord',\n",
        " 'namespace': u'org.bdgenomics.adam.avro',\n",
        " 'type': u'record'}"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMGenotype"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMContig\n",
      "\n",
      "We'll start with the ADAMContig record, which is used to describe regions of DNA that has overlapping reads."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "*To save space for future writes and reads of records, which we'll run in-notebook to create example records and verify that our records are stored properly, we'll use the below function.*"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMVariant\n",
      "\n",
      "An ADAMVariant represents a variant (with respect to the reference allele) detected at a given position in a contig.\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "scm = protocol.types_dict['ADAMVariant']\n",
      "scm.to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ADAM_variant_1 = {'position': 9882L, 'referenceAllele': \"A\", 'variantAllele': \"T\", 'contig': ADAM_contig_1}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "write_then_read_records(scm, 'variants.avro', [ADAM_variant_1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMNucleotideContigFragment\n",
      "\n",
      "Next we'll look at an ADAMNucleotideContigFragment, which\u2026\u00a0\n",
      "\n",
      "> stores a contig of nucleotides; this may be a reference chromosome, may be an assembly, may be a BAC. very long contigs (>1Mbp) need to be split into fragments. it seems that they are too long to load in a single go.\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "scm = protocol.types_dict['ADAMNucleotideContigFragment']\n",
      "scm.to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ADAM_nucleotide_contig_frag = {'contigName': \"Jimmy\", 'contigID': 9647, 'description': \"A solid contig, a real workhorse\", \n",
      "                               'url': \"http://contigs.org/9647\", 'fragmentLength': 203,\n",
      "                               'contigLength': 767, 'fragmentNumber': 2, 'fragmentStartPosition': 7789,\n",
      "                               'numberOfFragmentsInContig': 17}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "write_then_read_records(scm, 'nuc_contig_frag.avro', [ADAM_nucleotide_contig_frag])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMPileup\n",
      "\n",
      "ADAMPileup summarizes base calls against a reference sequence of aligned reads with various coverage."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "scm = protocol.types_dict['ADAMPileup']\n",
      "scm.to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ADAM_pileup = {'referenceName': \"Tony\", \"referenceID\": 7837, 'position': 443, 'rangeOffset': 22, 'rangeLength': 778,\n",
      "               'referenceBase': 'C', 'readBase': 'C', 'sangerQuality': 22, 'mapQuality': 19, 'numSoftClipped': 2,\n",
      "               'numReverseStrand': 1, 'countAtPosition': 1, 'readName': \"XYZ\", 'readStart': 223L, 'readEnd': 298L,\n",
      "               'recordGroupSequencingCenter': \"...\", 'recordGroupDescription': \"...\", 'recordGroupRunDateEpoch': 1101028284,\n",
      "               'recordGroupFlowOrder': \"forward\", 'recordGroupKeySequence': \"...\", 'recordGroupLibrary': \"...\",\n",
      "               'recordGroupPredictedMedianInsertSize': 2, 'recordGroupPlatform': \"...\", 'recordGroupPlatformUnit': \"...\",\n",
      "               'recordGroupSample': \"...\"}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "write_then_read_records(scm, 'pileup.avro', [ADAM_pileup])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### VariantEffect\n",
      "\n",
      "VariantEffect denotes the effect that a variant allele has on a given gene. It notes the change, if any, in amino acid for a given codon's mutation."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "scm = protocol.types_dict['VariantEffect']\n",
      "scm.to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "variant_effect = {'referenceAminoAcid': \"PRO\", 'alternateAminoAcid': \"SER\", 'geneId': 'MYO6', 'transcriptId': \"MYO6-001\"}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "write_then_read_records(scm, 'var_effect.avro', [variant_effect])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### VariantCallingAnnotations"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "3scm = protocol.types_dict['VariantCallingAnnotations']\n",
      "scm.to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "variant_calling_annotations = {'readDepth': 60, 'downSampled': False, 'baseQRankSum': 34.7, 'clippingRankRum': 99.2,\n",
      "                               'fisherStrandBiasPValue': 33.3, 'haplotypeScore': 22.7, 'inbreedingCoefficient': 2.0,\n",
      "                               'alleleCountMLE': [5,1,5,3,2,3,4,9], 'alleleFrequencyMLE': [5,1,5,3,2,3,4,9], 'rmsMapQ': 12.7,\n",
      "                               'mapq0Reads': 19, 'mpRankSum': 276.9, 'usedForNegativeTrainingSet': False, \n",
      "                               'usedForPositiveTrainingSet': True, 'variantQualityByDepth': 11.4, 'readPositionRankSum': 12.3,\n",
      "                               'vqslod': 1.0, 'culprit': \"A\", 'variantCallErrorProbability': 0.14, 'variantIsPassing': True,\n",
      "                               'variantFilters': ['ags', 'mqr']}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "write_then_read_records(scm, 'var_calling_anno.avro', [variant_calling_annotations])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMDatabaseVariantAnnotation"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### ADAMNestedPileup\n",
      "\n",
      "> nested pileup data type - contains reference to list of overlapping reads  \n",
      "> note: cannot be used with databases (e.g. hive/shark)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "scm = protocol.types_dict['ADAMNestedPileup']\n",
      "scm.to_json()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ADAM_nested_pileup = {'pileup': ADAM_pileup, 'readEvidence': [ADAM_record]}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}